File: src/main/java/upbrella/be/store/entity/Classification.java
package upbrella.be.store.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import upbrella.be.store.dto.request.CreateClassificationRequest;
import upbrella.be.store.dto.request.CreateSubClassificationRequest;

import javax.persistence.*;

@Entity
@Getter
@Builder
@NoArgsConstructor(access = lombok.AccessLevel.PROTECTED)
@AllArgsConstructor
public class Classification {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;
    @Enumerated(EnumType.STRING)
    private ClassificationType type;
    private String name;
    private Double latitude;
    private Double longitude;

    public static Classification ofCreateClassification(CreateClassificationRequest request) {

        return Classification.builder()
                .name(request.getName())
                .type(ClassificationType.CLASSIFICATION)
                .latitude(request.getLatitude())
                .longitude(request.getLongitude())
                .build();
    }

    public static Classification ofCreateSubClassification(CreateSubClassificationRequest request) {

        return Classification.builder()
                .name(request.getName())
                .type(ClassificationType.SUB_CLASSIFICATION)
                .build();
    }
}

File: src/main/kotlin/upbrella/be/store/entity/Classification.kt
package upbrella.be.store.entity

import upbrella.be.store.dto.request.CreateClassificationRequest
import upbrella.be.store.dto.request.CreateSubClassificationRequest
import javax.persistence.*

@Entity
class Classification(
    @Enumerated(EnumType.STRING)
    val type: ClassificationType? = null,
    val name: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,
) {
    companion object {
        @JvmStatic
        fun ofCreateClassification(request: CreateClassificationRequest): Classification {
            return Classification(
                type = ClassificationType.CLASSIFICATION,
                name = request.name,
                latitude = request.latitude,
                longitude = request.longitude,
            )
        }

        @JvmStatic
        fun ofCreateSubClassification(request: CreateSubClassificationRequest): Classification {
            return Classification(
                type = ClassificationType.SUB_CLASSIFICATION,
                name = request.name,
            )
        }
    }
}